spec/Subscription-spec.ts

```diff
@@ -200,5 +200,28 @@ describe('Subscription', () => {
       subscription.unsubscribe();
       expect(count).to.equal(1);
     });
+
+    it('should unsubscribe from all parents', () => {
+      // https://github.com/ReactiveX/rxjs/issues/6351
+      const a = new Subscription(() => { /* noop */});
+      const b = new Subscription(() => { /* noop */});
+      const c = new Subscription(() => { /* noop */});
+      const d = new Subscription(() => { /* noop */});
+      a.add(d);
+      b.add(d);
+      c.add(d);
+      // When d is added to the subscriptions, it's added as a teardown. The
+      // length is 1 because the teardowns passed to the ctors are stored in a
+      // separate property.
+      expect((a as any)._teardowns).to.have.length(1);
+      expect((b as any)._teardowns).to.have.length(1);
+      expect((c as any)._teardowns).to.have.length(1);
+      d.unsubscribe();
+      // When d is unsubscribed, it should remove itself from each of its
+      // parents.
+      expect((a as any)._teardowns).to.have.length(0);
+      expect((b as any)._teardowns).to.have.length(0);
+      expect((c as any)._teardowns).to.have.length(0);
```

Comment:
I could re-write this test as a memory leak test - using `FinalizationRegistry` instead of the private `_teardown` property - but IDK. It's fairly straightforward ATM.